File: smoke_tests/swift/regex_dos.swift
// Regular Expression Denial of Service (ReDoS) vulnerabilities in Swift
import Foundation

class RegexDosVulnerabilities {

    // Test 1: Catastrophic backtracking with nested quantifiers
    func validateEmail(email: String) -> Bool {
        // VULNERABLE: Nested quantifiers cause exponential backtracking
        let pattern = "^([a-zA-Z0-9]+)+@[a-zA-Z0-9]+\\.[a-zA-Z]+$"
        return email.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 2: Overlapping alternation
    func validateInput(input: String) -> Bool {
        // VULNERABLE: Overlapping patterns
        let pattern = "^(a+|a+b)+$"
        return input.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 3: User-provided regex
    func matchPattern(input: String, pattern: String) -> Bool {
        // VULNERABLE: User controls regex pattern
        return input.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 4: Complex URL validation
    func validateUrl(url: String) -> Bool {
        // VULNERABLE: Complex regex with backtracking
        let pattern = "^(https?://)?([\\da-z\\.-]+)\\.([a-z\\.]{2,6})([/\\w \\.-]*)*/?$"
        return url.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 5: HTML tag stripping
    func stripHtml(html: String) -> String {
        // VULNERABLE: Greedy quantifiers on user input
        let pattern = "<[^>]*>"
        return html.replacingOccurrences(of: pattern, with: "", options: .regularExpression)
    }

    // Test 6: Password strength check
    func isStrongPassword(password: String) -> Bool {
        // VULNERABLE: Multiple lookaheads with quantifiers
        let pattern = "^(?=.*[a-z]+)(?=.*[A-Z]+)(?=.*[0-9]+)(?=.*[!@#$%^&*]+).{8,}$"
        return password.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 7: Log parsing regex
    func parseLogLine(line: String) -> [String]? {
        // VULNERABLE: Greedy .* patterns
        let pattern = "^(.*) - (.*) \\[(.*?)\\] \"(.*)\" (\\d+) (.*)$"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(line.startIndex..., in: line)
        guard let match = regex.firstMatch(in: line, range: range) else { return nil }
        return (1..<match.numberOfRanges).compactMap { i in
            Range(match.range(at: i), in: line).map { String(line[$0]) }
        }
    }

    // Test 8: Repeated groups
    func validateFormat(input: String) -> Bool {
        // VULNERABLE: Repeated capturing groups
        let pattern = "^(([a-z])+\\.)+[a-z]+$"
        return input.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 9: Whitespace normalization
    func normalizeWhitespace(text: String) -> String {
        // VULNERABLE: Multiple spaces pattern on large input
        let pattern = "\\s+"
        return text.replacingOccurrences(of: pattern, with: " ", options: .regularExpression)
    }

    // Test 10: CSV field extraction
    func extractFields(csvLine: String) -> [String] {
        // VULNERABLE: Complex CSV regex
        let pattern = "(?:^|,)(?:\"([^\"]*(?:\"\"[^\"]*)*)\"|([^\",]*))"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        let range = NSRange(csvLine.startIndex..., in: csvLine)
        let matches = regex.matches(in: csvLine, range: range)
        return matches.compactMap { match in
            Range(match.range, in: csvLine).map { String(csvLine[$0]) }
        }
    }

    // Test 11: NSRegularExpression with timeout
    func matchWithTimeout(input: String, pattern: String) -> Bool {
        // VULNERABLE: No timeout mechanism
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
        let range = NSRange(input.startIndex..., in: input)
        return regex.firstMatch(in: input, range: range) != nil
    }

    // Test 12: Phone number validation
    func validatePhone(phone: String) -> Bool {
        // VULNERABLE: Alternation with backtracking
        let pattern = "^(\\+\\d{1,3}[- ]?)?(\\(\\d{1,4}\\)[- ]?)?\\d{1,4}([- ]?\\d{1,4}){1,3}$"
        return phone.range(of: pattern, options: .regularExpression) != nil
    }

    // Test 13: JSON-like validation
    func validateJsonLike(input: String) -> Bool {
        // VULNERABLE: Recursive-like patterns
        let pattern = "^\\{(\"[^\"]+\":\\s*(\"[^\"]*\"|\\d+|true|false|null),?\\s*)+\\}$"
        return input.range(of: pattern, options: .regularExpression) != nil
    }
}

File: smoke_tests/dart/regex_dos.dart
// Regular Expression Denial of Service (ReDoS) vulnerabilities in Dart

// Test 1: Evil regex with nested quantifiers
bool vulnerableNestedQuantifiers(String input) {
  // VULNERABLE: Nested quantifiers - exponential backtracking
  var regex = RegExp(r'^(a+)+$');
  return regex.hasMatch(input);
}

// Test 2: Overlapping alternatives
bool vulnerableOverlapping(String input) {
  // VULNERABLE: Overlapping patterns
  var regex = RegExp(r'^(a|a)+$');
  return regex.hasMatch(input);
}

// Test 3: Greedy with backtracking
bool vulnerableGreedy(String input) {
  // VULNERABLE: Greedy pattern with backtracking
  var regex = RegExp(r'^.*.*.*$');
  return regex.hasMatch(input);
}

// Test 4: User-controlled regex
bool vulnerableUserRegex(String pattern, String input) {
  // VULNERABLE: User-controlled regex pattern
  var regex = RegExp(pattern);
  return regex.hasMatch(input);
}

// Test 5: Email regex (common vulnerable pattern)
bool vulnerableEmailRegex(String email) {
  // VULNERABLE: Complex email regex with backtracking
  var regex = RegExp(r'^([a-zA-Z0-9]+([._-][a-zA-Z0-9]+)*)+@([a-zA-Z0-9]+([.-][a-zA-Z0-9]+)*)+$');
  return regex.hasMatch(email);
}

// Test 6: URL regex
bool vulnerableUrlRegex(String url) {
  // VULNERABLE: URL regex with potential backtracking
  var regex = RegExp(r'^(https?://)?([a-zA-Z0-9-]+\.)+[a-zA-Z]+(/[a-zA-Z0-9-._~:/?#\[\]@!$&\'()*+,;=%]*)*$');
  return regex.hasMatch(url);
}

// Test 7: HTML tag matching
String vulnerableHtmlStrip(String html) {
  // VULNERABLE: HTML regex with backtracking
  var regex = RegExp(r'<[^>]*>*');
  return html.replaceAll(regex, '');
}

// Test 8: Repeated group with alternatives
bool vulnerableRepeatedGroup(String input) {
  // VULNERABLE: Repeated group with alternatives
  var regex = RegExp(r'^([a-z]|[A-Z]|[0-9])+$');
  return regex.hasMatch(input);
}

// Test 9: Nested capturing groups
List<String?> vulnerableNestedCapture(String input) {
  // VULNERABLE: Nested capturing groups
  var regex = RegExp(r'^((a+)(b+))+$');
  var match = regex.firstMatch(input);
  return match?.groups([1, 2, 3]) ?? [];
}

// Test 10: Catastrophic backtracking
bool vulnerableCatastrophic(String input) {
  // VULNERABLE: Classic catastrophic backtracking
  var regex = RegExp(r'^(a+)+b$');
  return regex.hasMatch(input);
}

// Test 11: Optional with repetition
bool vulnerableOptionalRepeat(String input) {
  // VULNERABLE: Optional group with repetition
  var regex = RegExp(r'^(a?a?)+$');
  return regex.hasMatch(input);
}

// Test 12: Alternation with overlap
bool vulnerableAlternation(String input) {
  // VULNERABLE: Alternation with overlapping patterns
  var regex = RegExp(r'^(ab|a)+$');
  return regex.hasMatch(input);
}

// Test 13: Complex pattern from config
bool vulnerableConfigPattern(String configPattern, String input) {
  // VULNERABLE: Pattern from configuration
  try {
    var regex = RegExp(configPattern);
    return regex.hasMatch(input);
  } catch (e) {
    return false;
  }
}

// Test 14: Search with user pattern
List<String> vulnerableSearch(String pattern, List<String> items) {
  // VULNERABLE: User pattern in search
  var regex = RegExp(pattern);
  return items.where((item) => regex.hasMatch(item)).toList();
}

// Test 15: Split with user regex
List<String> vulnerableSplit(String input, String pattern) {
  // VULNERABLE: User pattern in split
  return input.split(RegExp(pattern));
}

// Test 16: Replace with user regex
String vulnerableReplace(String input, String pattern, String replacement) {
  // VULNERABLE: User pattern in replace
  return input.replaceAll(RegExp(pattern), replacement);
}

// Test 17: Quadratic blowup
bool vulnerableQuadratic(String input) {
  // VULNERABLE: Quadratic complexity
  var regex = RegExp(r'^(.*?,)+.*$');
  return regex.hasMatch(input);
}

// Test 18: Match all with vulnerable pattern
Iterable<RegExpMatch> vulnerableMatchAll(String input) {
  // VULNERABLE: Match all with backtracking pattern
  var regex = RegExp(r'(a+)+');
  return regex.allMatches(input);
}
